### ISSUE TITLE
Add Benchmark

### ISSUE THREAD

**Issue by hishamco** [hishamco/My.Extensions.Localization.Json#37]
Add benchmark project to measure the performance

**Comment by fededim** [comment]
It would be better because I noticed that the localization of a string takes around 250ms which is extremely slow (I have a JSON with 2600 keys)! The problem is in the string localizer where you use a Dictionary only for indexing culture and not keys (it's an enumerable which has be scanned for the search of each key, if you have a lot of keys it's very slow). It would be better to use a double dictionary, e.g.

Change from

private readonly ConcurrentDictionary<string, IEnumerable<KeyValuePair<string, string>>> _resourcesCache = new ConcurrentDictionary<string, IEnumerable<KeyValuePair<string, string>>>();

To 

private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, string>> _resourcesCache = new  ConcurrentDictionary<string, ConcurrentDictionary<string, string>>();

Moreover the cache has to be built only once at the beginning and not at every keyword search.

**Comment by hishamco** [comment]
Thanks @fededim for reporting this, just one question did the PERF improved after you did your changes?

I will try to finish the benchmarking and modify some places in the code to improve the PERF

**Comment by fededim** [comment]
I ended up writing my own library which takes around 0ms (so almost 200 times less). The speed greatly improves because for an access Dictionary has O(1) time complexity (e.g. constant in the dictionary size) whereas IEnumerable has O(n) time complexity (e.g. linear in the number of elements of IEnumerable).

**Comment by hishamco** [comment]
It would be nice to push a PR to improve the PERF

**Comment by hishamco** [comment]
> Moreover the cache has to be built only once at the beginning and not at every keyword search.

As I remembered I did this to make the localization keys updated whenever I modify the json file, probably I need to revise everything in there

Thanks

**Comment by fededim** [comment]
I do not have time for a PR (I have already written my own library), the changes are not difficult just change the variable I pointed out and fix compilation errors. It is just an advice of how to improve the speed of your library.

**Comment by hishamco** [comment]
Thanks @fededim I will try to measure the PERF after your changes

**Comment by fededim** [comment]
> > Moreover the cache has to be built only once at the beginning and not at every keyword search.
> 
> As I remembered I did this to make the localization keys updated whenever I modify the json file, probably I need to revise everything in there
> 
> Thanks

Ok this is a good point, but do you need it ? Usually localization happens rarely: when you support a new language you translate all strings and create a new json file which remains more or less the same. There might be some corrections, but in that case you can change the file and restart the website/application pool to reload the new changes.
If you really want an autorefresh the best way would be to provide a InvalidateCache method and set a watcher on the resource directory which calls the InvalidateCache method whenever any file changes. The string operator of IStringLocalizer should reload the cache only if it has been invalidated.

**Comment by hishamco** [comment]
FYI configuration APIs is already have auto reload options buitl-in

**Comment by fededim** [comment]
Fine, but you need still a callback to update the Dictionary when json files are reloaded (maybe you can use directly the IConfigurationRoot returned from Build command instead of using a Dictionary, but I do not know its performance). In this case you also need synchronized Dictionary (like the ConcurrentDictionary you use) whereas if you don't use autorefresh you can use standard Dictionary which is faster. IMHO autorefresh is not needed for Localization.

**Comment by hishamco** [comment]
>  IMHO autorefresh is not needed for Localization.

Yes & No, it's needed to refresh the added values at runtime, but not for a new added culture

**Comment by fededim** [comment]
I checked [ConfigurationRoot](https://github.com/aspnet/Configuration/blob/master/src/Config/ConfigurationRoot.cs), every [provider](https://github.com/aspnet/Configuration/blob/d469707ab18eef7ed0002f00175a9ad5b0f36250/src/Config/ConfigurationProvider.cs) uses a dictionary to store the key-values, even though it is not synchronized (this could raise the issue of not finding a key during a reload). If you want you can use also a Dictionary<string, ConfigurationRoot> exploiting its autoreload.

**Comment by hishamco** [comment]
Thanks for your informative details, I will try to made some changes ASAP

**Comment by fededim** [comment]
You are welcome

**Comment by hishamco** [comment]
@fededim could you please share your json file if you don't mind, I need to measure the PERF one more time

**Comment by hishamco** [comment]
Fixed in #74
